openedx/openedx-platform
# [BD-13] deprecates runtime.xqueue in favor of the XQueueService

## Description

Builds on the shim added by https://github.com/edx/edx-platform/pull/29190 to deprecate the `ModuleSystem.xqueue` property, which is only used by CAPA problems.

This change is only a refactoring, and should not affect Learners, Course Authors, or anyone else using edx-platform.

## Supporting information

* [Content Core Platform Simplication project](https://openedx.atlassian.net/wiki/spaces/AC/pages/3091629503/Content+Core+Platform+Simplification)
* [BLENDED-113](https://openedx.atlassian.net/browse/BLENDED-113) (edX internal ticket)
* [FAL-2209](https://tasks.opencraft.com/browse/FAL-2209) (OpenCraft internal ticket)

## Testing instructions

* LMS: https://pr29312.sandbox.opencraft.hosting/
* Studio: https://studio.pr29312.sandbox.opencraft.hosting/

LMS:
1. Login as a demo learner (`honor@example.com` / `edx`)
2. Enroll in the BD-13 test course
3. Navigate to [Complex blocks > External grader](https://discovery.pr29312.sandbox.opencraft.hosting/learning/course/course-v1:OpenCraft+BD-13+2021_1/block-v1:OpenCraft+BD-13+2021_1+type@sequential+block@862667c509144a4c83381589185f9072/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@b7e89f99b71d44f9b88dd1775628f539)
4. Ensure that you can submit answers to the externally graded problem.

    ⚠️ Note: the grader is dumb, and so always returns "correct", so it doesn't matter what you submit.
5. Repeat this test in the legacy course experience.

Studio:
1. Login as a demo staff (`staff@example.com` / `edx`)
2. Navigate to the [BD-13 test course > Complex blocks > External grader](https://studio.pr29312.sandbox.opencraft.hosting/container/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@b7e89f99b71d44f9b88dd1775628f539)
3. Try submitting answers to the problem.

    You'll get a message that reads "Error: No grader has been set up for this problem. " because externally graded problems can only be graded in preview / LMS mode.

     You can verify this behaviour on the sandbox created for https://github.com/edx/edx-platform/pull/29260.
4. View the problem in Preview mode, and ensure you can submit answers to the problem as you can in the LMS.

**Regression testing:**

LMS:
1. Login as a demo learner (`honor@example.com` / `edx`)
2. Browse around the Demo course, paying particular attention to the CAPA problems.
3. Check that you can submit answers to problems, and save changes to blocks.
3. Switch to the Legacy Course Experience and ensure you can submit answers to problems there too.

Studio:
1. Login as a demo staff (`staff@example.com` / `edx`)
2. Navigate to the Demo course and browse through the blocks, paying particular attention to the CAPA problem blocks. Ensure they render and edit as expected.
3. Check that you can submit answers to problems, and save changes to blocks.
4. Preview unpublished changes and ensure they render as expected.
4. Publish changes, and ensure they render in the LMS as expected.

## Deadline

None

## Author Notes & Concerns

1. See [this comment](https://github.com/edx/edx-platform/pull/29312#pullrequestreview-814434914) for question about `xqueue_callback_url_prefix`.

Notes for <strike>hacking</strike> setting up the dummy external grader:

```bash
# on the sandbox appserver:
sudo mkdir /edx/app/grader
sudo chown pomegranited: /edx/app/grader
cd /edx/app/grader

# set up virtualenv
sudo apt install python3.8-venv
python3 -m venv venv/grader

# check out my dummy grader branch
git clone https://github.com/open-craft/xqueue-watcher/
cd xqueue-watcher
git checkout jill/bd-13-dummy-grader

# Run the grader in a screen session
screen -S grader
source /edx/app/grader/venv/grader/bin/activate
make requirements  # had to run this twice to get around some error with building a wheel for dogstatsd-python
python -m xqueue_watcher -d .
# suspend screen session
```
